objects.js

```diff
@@ -2634,100 +2634,164 @@ SpriteMorph.prototype.searchBlocks = function (
-                    if (!inputs[1].length) {
-                        inputs[1] = ch;
-                    } else {
-                        return parseInfix(
-                            expression.slice(idx),
-                            ch,
-                            [operator, already, format(inputs[1])]
-                        );
+            } else if (token.type === "identifier") {
+                if (tokens.length && tokens[0].type === "lp") {
+                    var args = [];
+                    tokens.shift();
+                    while (tokens[0].type !== "rp") {
+                        args.push(createASTFromTokens(tokens));
+                        if (tokens[0].type === "rp") {
+                            break;
+                        }
+                        if (tokens[0].type !== "comma") {
+                            throw "No comma between arguments";
+                        }
+                        tokens.shift();
                     }
+                    tokens.shift();
+                    return [token.val].concat(args);
                 } else {
-                    operator = ch;
-                    already = format(inputs[0]);
+                    return token.val;
+                }
+            } else {
+                throw "Invalid expression";
+            }
+        }
+        function precedence1(tokens) {
+            var exp = precedence0(tokens), op;
+            if (!tokens.length) {
+                return exp;
+            }
+            while (tokens.length && /[*/%]/.test(tokens[0].val)) {
+                op = tokens[0];
+                if (op.type !== "operator") {
+                    return exp;
+                }
+                switch (op.val) {
+                    case "*":
+                    case "/":
+                    case "%": break;
+                    default: return exp;
                 }
-                break;
-            default:
-                inputs[operator ? 1 : 0] += ch;
+                tokens.shift();
+                exp = [op.val, exp, precedence0(tokens)];
             }
+            return exp;
         }
-        if (operator) {
-            return [operator, already, format(inputs[1])];
+        function precedence2(tokens) {
+            var exp = precedence1(tokens), op;
+            if (!tokens) {
+                return exp;
+            }
+            while (tokens.length && /[+\-]/.test(tokens[0].val)) {
+                op = tokens[0];
+                if (op.type !== "operator") {
+                    return exp;
+                }
+                switch (op.val) {
+                    case "+":
+                    case "-": break;
+                    default: return exp;
+                }
+                tokens.shift();
+                exp = [op.val, exp, precedence1(tokens)];
+            }
+            return exp;
         }
-        return format(inputs[0]);
     }
-
     function blockFromAST(ast) {
```

Comment:
Oh, there's a question.  People who program in infix expect && and || for the Boolean operators, so you should certainly accept them (so that copy-and-paste works).  It might be okay to accept & and | _also._  On the other hand, we really should add bitand and bitor to the primitive dyadic operators someday.